js/piwik.js

```diff
@@ -7706,6 +7712,32 @@ if (typeof window.Piwik !== 'object') {
         // initialize the Piwik singleton
         addEventListener(windowAlias, 'beforeunload', beforeUnloadHandler, false);
 
+        window.addEventListener('message', function(e) {
+            try {
+                var data = JSON.parse(e.data);
```

Comment:
haven't tested it yet... do we need to check if the message is for us just to avoid any issues with potentially other messages? like include some 'matomo_optout' in the string and then check if `e.data && typeof e.data === 'string' && e.data.indexOf('matomo_optout') >= -1`? (not tested that particular code... and would need to make sure we post this string in the opt out)

Reply:
Have added some namespacing of the message attributes to make it less likely that we'll accidentally pick up messages intended for other scripts.  The best practice would be to also check the message origin against the domain we expect to receive it from, but I can't find a way to do this since the tracker and the tracked site could be on any domain.
I don't think I need to explicitly check that `e.data` is defined, is a string etc as I am catching the exceptions from `JSON.parse()` in situations like this.